### PR TITLE
migrate v1 aave & compound strategies

### PR DESCRIPTION
**Pull request by hezhihua81**

### CODE REVIEW

File: contracts/strategies/aave/StrategyAaveLendingPool.sol [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            );
+
+            // Deposit supplying token to Aave Lending Pool and mint aToken.
+            uint256 obtainedSupplyTokenAmount = IERC20(supplyToken).balanceOf(address(this));
+            IERC20(supplyToken).safeIncreaseAllowance(lendingPool, obtainedSupplyTokenAmount);
+            ILendingPool(lendingPool).deposit(supplyToken, obtainedSupplyTokenAmount, address(this), 0);
+        }
+
+        // sync the asset balance
+        assetAmount = IAToken(aToken).balanceOf(address(this));
```
Comment (Dominator008) [inline contracts/strategies/aave/StrategyAaveLendingPool.sol:205]: I don't think we need to store and keep track of `assetAmount`. `syncPrice` should query it.

Comment (hezhihua81) [reply contracts/strategies/aave/StrategyAaveLendingPool.sol]: updated

File: contracts/strategies/aave/StrategyAaveLendingPool.sol [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        }
+        require(sharesFromBuy >= _minSharesFromBuy, "failed min shares from buy");
+        require(amountFromSell >= _minAmountFromSell, "failed min amount from sell");
+        if (_buyAmount > amountFromSell) {
+            _buy(_buyAmount - amountFromSell);
+        } else if (_buyAmount < amountFromSell) {
+            _sell(amountFromSell - _buyAmount);
+        }
+
+        if (_buyAmount > 0) {
```
Comment (Dominator008) [inline contracts/strategies/aave/StrategyAaveLendingPool.sol:119]: I think this check should be based on `_sharesFromBuy`. Only `Buy` or `Sell` should be emitted, not both.

File: contracts/strategies/aave/StrategyAaveLendingPool.sol [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            shares = _buyAmount;
+            assetAmount = _buyAmount;
+            sharesFromBuy = _buyAmount;
+        } else {
+            sharesFromBuy = _buyAmount.mul(shares).div(assetAmount);
+            amountFromSell = _sellShares.mul(assetAmount).div(shares);
+            assetAmount = assetAmount.add(_buyAmount).sub(amountFromSell);
+            shares = shares.add(sharesFromBuy).sub(_sellShares);
+        }
+        require(sharesFromBuy >= _minSharesFromBuy, "failed min shares from buy");
```
Comment (Dominator008) [inline contracts/strategies/aave/StrategyAaveLendingPool.sol:111]: These checks should happen after the interaction with Aave, not before.

File: contracts/strategies/aave/StrategyAaveLendingPool.sol [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+import "../interfaces/aave/IStakedAave.sol";
+import "../interfaces/uniswap/IUniswapV2.sol";
+
+/**
+ * Deposits ERC20 token into Aave Lending Pool and issues stAaveLendingToken(e.g. stAaveLendingDAI) in L2. Holds aToken (Aave interest-bearing tokens).
+ */
+contract StrategyAaveLendingPool is IStrategy, Ownable {
+    using SafeERC20 for IERC20;
+    using Address for address;
+    using SafeMath for uint256;
```
[the remark this answers is not in the document]
Comment (hezhihua81) [reply contracts/strategies/aave/StrategyAaveLendingPool.sol]: updated

File: contracts/strategies/aave/StrategyAaveLendingPool.sol [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        // 1. Deposit or withdrawal
+        uint256 sharesFromBuy;
+        uint256 amountFromSell;
+        uint256 assetAmount = IAToken(aToken).balanceOf(address(this));
+        if (assetAmount == 0 || shares == 0) {
+            shares = _buyAmount;
+            assetAmount = _buyAmount;
+            sharesFromBuy = _buyAmount;
+        } else {
+            sharesFromBuy = _buyAmount.mul(shares).div(assetAmount);
```
[the remark this answers is not in the document]
Comment (hezhihua81) [reply contracts/strategies/aave/StrategyAaveLendingPool.sol]: updated

File: contracts/strategies/aave/StrategyAaveLendingPool.sol (hunk 2 above)
Comment (hezhihua81) [reply contracts/strategies/aave/StrategyAaveLendingPool.sol]: updated

File: contracts/strategies/aave/StrategyAaveLendingPool.sol (hunk 3 above)
Comment (hezhihua81) [reply contracts/strategies/aave/StrategyAaveLendingPool.sol:109]: not an issue for aave

**Review by Dominator008 — APPROVED** [review]

**Pull request merged**